src/core/thread.d

```diff
@@ -631,7 +633,8 @@ class Thread
         }
 
         // Start thread as non-suspendable, undone in thread_entryPoint
-        atomicStore(m_isInCriticalRegion, true);
+        synchronized (Thread.criticalRegionLock)
+            m_isInCriticalRegion = true;
```

Comment:
> So thread_suspendAll wouldn't work when being called from DllMain. As said before we should do as little as possible in DllMain.

Not sure we can disallow allocating, which can always trigger a collection.

> I'll work out a different solution anyhow, the real problem is, that we're adding threads to the global list before they are started.

IIRC the thread had to be on the global list so that DllMain could detect that the thread is actually created by the runtime in the DLL_THREAD_ATTACH callback. Maybe this can be solved differently.